deepchecks/base/check.py

```diff
@@ -431,10 +445,53 @@ def run(self,
         if len(test_datasets) != len(models):
             raise DeepchecksValueError('number of test_datasets must equal to number of models')
 
-        return self._run(list(zip(train_datasets, test_datasets, models)))
+        self.train_datasets = train_datasets
+        self.test_datasets = test_datasets
+        self.model_names = list(models.keys())
+        self.models = list(models.values())
+
+        # Additional validations
+        self.task_type = None
+        for i in range(len(models)):
+            train = self.train_datasets[i]
+            test = self.test_datasets[i]
+            model = self.models[i]
+            Dataset.validate_dataset(train)
+            Dataset.validate_dataset(test)
+            train.validate_label()
+            train.validate_features()
+            train.validate_shared_features(test)
+            train.validate_shared_label(test)
+            validate_model(train, model)
+            curr_task_type = task_type_check(model, train)
+            if self.task_type is None:
+                self.task_type = curr_task_type
+            elif curr_task_type != self.task_type:
+                raise DeepchecksNotSupportedError('Got models of different task types')
+
+    def __len__(self):
+        """Return number of models."""
+        return len(self.models)
+
+    def __iter__(self):
+        """Return iterator over context objects."""
+        return zip(self.train_datasets, self.test_datasets, self.models, self.model_names)
+
+
+class ModelComparisonBaseCheck(BaseCheck):
+    """Parent class for check that compares between two or more models."""
+
+    def run(self,
+            train_datasets: Union[Dataset, List[Dataset]],
+            test_datasets: Union[Dataset, List[Dataset]],
+            models: Union[List[Any], Mapping[str, Any]]
+            ) -> CheckResult:
+        """Initialize context and pass to check logic."""
+        return self.run_logic(ModelComparisonContext(train_datasets, test_datasets, models))
 
     @abc.abstractmethod
-    def _run(self, datasets_models: List[Tuple[Dataset, Dataset, Any]]):
+    def run_logic(self, context: ModelComparisonContext):
```

Comment:
Maybe "run_with_context" or "run_model_comparison_internal" ?

Reply:
that's ugly name, and this is the method that will be implemented in each check, so it can't be too ugly